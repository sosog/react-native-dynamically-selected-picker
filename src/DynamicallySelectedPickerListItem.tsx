import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import type { ListItem, PickerListItemProps } from './types/pickerTypes';

export default function DynamicallySelectedPickerListItem<
  ItemT extends ListItem
>({
  item,
  allItemsColor,
  fontSize,
  height,
  fontFamily = 'Arial',
}: PickerListItemProps<ItemT>) {
  return (
    <View
      style={{
        ...styles.viewWrapper,
        height: height,
      }}
    >
      <Text
        style={{
          fontSize: fontSize,
          color: item.itemColor ? item.itemColor : allItemsColor,
          fontFamily: fontFamily,
        }}
      >
        {item.label}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  viewWrapper: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});
