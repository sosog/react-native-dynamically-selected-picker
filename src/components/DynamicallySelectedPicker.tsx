import React, { createRef, useState } from 'react';

import {
  StyleSheet,
  View,
  ScrollView,
  Platform,
  NativeScrollEvent,
  NativeSyntheticEvent,
} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import DynamicallySelectedPickerListItem from '../DynamicallySelectedPickerListItem';
import type { ListItem, PickerProps } from '../types/pickerTypes';

export default function DynamicallySelectedPicker({
  items = [{ value: 0, label: 'No items', itemColor: 'red' }],
  onScroll,
  onScrollBeginDrag,
  onScrollEndDrag,
  onMomentumScrollBegin,
  onMomentumScrollEnd,
  width = 300,
  height = 300,
  initialSelectedIndex = 0,
  transparentItemRows = 3,
  allItemsColor = '#000',
  fontFamily = 'Arial',
  fontSize,
  selectedItemBorderColor = '#cecece',
  renderGradientOverlay = true,
  topGradientColors = [
    'rgba( 255, 255, 255, 1 )',
    'rgba( 255, 255, 255, 0.9 )',
    'rgba( 255, 255, 255, 0.7 )',
    'rgba( 255, 255, 255, 0.5 )',
  ],
  bottomGradientColors = [
    'rgba( 255, 255, 255, 0.5 )',
    'rgba( 255, 255, 255, 0.7 )',
    'rgba( 255, 255, 255, 0.9 )',
    'rgba( 255, 255, 255, 1 )',
  ],
}: PickerProps) {
  let itemHeightInitial = height / (transparentItemRows * 2 + 1);
  if (Platform.OS === 'ios') {
    itemHeightInitial = Math.ceil(itemHeightInitial);
  }
  const [itemHeight] = useState<number>(itemHeightInitial);
  const [itemIndex, setItemIndex] = useState<number>(initialSelectedIndex);

  const scrollViewRef = createRef<ScrollView>();

  const scrollToInitialPosition = () => {
    scrollViewRef.current?.scrollTo({
      y: itemHeight * initialSelectedIndex,
    });
  };

  function fakeItems(n = 3) {
    const itemsArr = [];
    for (let i = 0; i < n; i++) {
      itemsArr[i] = {
        value: -1,
        label: '',
      };
    }
    return itemsArr;
  }

  function allItemsLength() {
    return extendedItems().length - transparentItemRows * 2;
  }

  function onScrollListener(event: NativeSyntheticEvent<NativeScrollEvent>) {
    if (onScroll != null) {
      const index = getItemIndex(event);
      if (itemIndex !== index && index >= 0 && index < allItemsLength()) {
        setItemIndex(index);
        onScroll({ index });
      }
    }
  }

  function onMomentumScrollBeginListener(
    event: NativeSyntheticEvent<NativeScrollEvent>
  ) {
    if (onMomentumScrollBegin != null) {
      const index = getItemIndex(event);
      if (index >= 0 && index < allItemsLength()) {
        setItemIndex(index);
        onMomentumScrollBegin({ index });
      }
    }
  }

  function onMomentumScrollEndListener(
    event: NativeSyntheticEvent<NativeScrollEvent>
  ) {
    if (onMomentumScrollEnd != null) {
      const index = getItemIndex(event);

      if (index >= 0 && index < allItemsLength()) {
        setItemIndex(index);
        onMomentumScrollEnd({ index });
      }
    }
  }

  function onScrollBeginDragListener(
    event: NativeSyntheticEvent<NativeScrollEvent>
  ) {
    if (onScrollBeginDrag != null) {
      const index = getItemIndex(event);

      if (index >= 0 && index < allItemsLength()) {
        setItemIndex(index);
        onScrollBeginDrag({ index });
      }
    }
  }

  function onScrollEndDragListener(
    event: NativeSyntheticEvent<NativeScrollEvent>
  ) {
    if (onScrollEndDrag != null) {
      const index = getItemIndex(event);

      if (index >= 0 && index < allItemsLength()) {
        setItemIndex(index);
        onScrollEndDrag({ index });
      }
    }
  }

  function getItemIndex(event: NativeSyntheticEvent<NativeScrollEvent>) {
    return Math.round(event.nativeEvent.contentOffset.y / itemHeight);
  }

  function extendedItems() {
    return [
      ...fakeItems(transparentItemRows),
      ...items,
      ...fakeItems(transparentItemRows),
    ];
  }

  const position = {
    top: 0,
    bottom: 0,
  };

  const border = {
    topWidth: 1,
    bottomWidth: 1,
  };

  return (
    <View style={{ height, width }}>
      <ScrollView
        ref={scrollViewRef}
        onLayout={scrollToInitialPosition}
        showsVerticalScrollIndicator={false}
        showsHorizontalScrollIndicator={false}
        onMomentumScrollBegin={onMomentumScrollBeginListener}
        onMomentumScrollEnd={onMomentumScrollEndListener}
        onScrollBeginDrag={onScrollBeginDragListener}
        onScrollEndDrag={onScrollEndDragListener}
        onScroll={onScrollListener}
        scrollEventThrottle={20}
        snapToInterval={itemHeight}
      >
        {extendedItems().map((item: ListItem, index) => {
          return (
            <DynamicallySelectedPickerListItem
              key={index}
              label={item.label}
              itemColor={item.itemColor}
              allItemsColor={allItemsColor}
              fontSize={fontSize ? fontSize : itemHeight / 2}
              fontFamily={fontFamily}
              height={itemHeight}
            />
          );
        })}
      </ScrollView>
      {renderGradientOverlay && (
        <LinearGradient
          colors={topGradientColors}
          style={[
            styles.gradientWrapper,
            {
              top: position.top,
              height: transparentItemRows * itemHeight,
            },
          ]}
          pointerEvents="none"
        />
      )}
      <View
        style={[
          styles.gradientWrapper,
          {
            top: transparentItemRows * itemHeight,
            borderBottomWidth: border.bottomWidth,
            borderBottomColor: selectedItemBorderColor,
          },
        ]}
        pointerEvents="none"
      />
      {renderGradientOverlay && (
        <LinearGradient
          colors={bottomGradientColors}
          style={[
            styles.gradientWrapper,
            {
              bottom: position.bottom,
              height: transparentItemRows * itemHeight,
            },
          ]}
          pointerEvents="none"
        />
      )}
      <View
        style={[
          styles.gradientWrapper,
          {
            bottom: transparentItemRows * itemHeight,
            borderTopWidth: border.topWidth,
            borderTopColor: selectedItemBorderColor,
          },
        ]}
        pointerEvents="none"
      />
    </View>
  );
}

const styles = StyleSheet.create({
  listItem: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  gradientWrapper: {
    position: 'absolute',
    width: '100%',
  },
});
