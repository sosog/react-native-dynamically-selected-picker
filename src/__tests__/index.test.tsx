import React from 'react';
import renderer from 'react-test-renderer';

import DynamicallySelectedPicker from '../components/DynamicallySelectedPicker';
import DynamicallySelectedPickerListItem from '../DynamicallySelectedPickerListItem';
import type { ListItem } from 'src/types/pickerTypes';

it('Components renders correctly with defaults', () => {
  const DynamicallySelectedPickerWrapper = renderer
    .create(<DynamicallySelectedPicker />)
    .toJSON();
  expect(DynamicallySelectedPickerWrapper).toMatchSnapshot();

  const item: ListItem = {
    label: 'Item',
    value: 1,
    itemColor: 'white',
  };

  const DynamicallySelectedPickerListItemWrapper = renderer
    .create(
      <DynamicallySelectedPickerListItem
        item={item}
        key={1}
        isSelected={false}
        allItemsColor={'black'}
        fontSize={14}
        height={30}
        fontFamily={'Georgia'}
      />
    )
    .toJSON();
  expect(DynamicallySelectedPickerListItemWrapper).toMatchSnapshot();
});

it('Components renders correctly with updated params', () => {
  const DynamicallySelectedPickerWrapper = renderer
    .create(
      <DynamicallySelectedPicker
        items={[
          {
            value: 1,
            label: 'Item 1',
          },
          {
            value: 2,
            label: 'Item 2',
          },
          {
            value: 3,
            label: 'Item 3',
          },
          {
            value: 4,
            label: 'Item 4',
            itemColor: 'blue',
          },
          {
            value: 5,
            label: 'Item 5',
          },
        ]}
        initialSelectedIndex={2}
        height={400}
        width={400}
        fontSize={18}
        allItemsColor={'blue'}
        fontFamily={'Arial'}
        selectedItemBorderColor={'gray'}
        transparentItemRows={2}
      />
    )
    .toJSON();
  expect(DynamicallySelectedPickerWrapper).toMatchSnapshot();

  const item: ListItem = {
    label: 'Item label',
    value: 1,
    itemColor: 'green',
  };

  const DynamicallySelectedPickerListItemWrapper = renderer
    .create(
      <DynamicallySelectedPickerListItem
        item={item}
        key={1}
        isSelected={false}
        allItemsColor={'pink'}
        fontSize={22}
        height={40}
        fontFamily={'Arial'}
      />
    )
    .toJSON();
  expect(DynamicallySelectedPickerListItemWrapper).toMatchSnapshot();
});

it('Components renders correctly with gradient disabled', () => {
  const DynamicallySelectedPickerWrapper = renderer
    .create(<DynamicallySelectedPicker renderGradientOverlay={false} />)
    .toJSON();
  expect(DynamicallySelectedPickerWrapper).toMatchSnapshot();
});
