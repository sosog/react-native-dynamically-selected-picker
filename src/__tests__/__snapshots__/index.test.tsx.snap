// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components renders correctly with defaults 1`] = `
<View
  style={
    {
      "height": 300,
      "width": 300,
    }
  }
>
  <RCTScrollView
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    scrollEventThrottle={20}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    snapToInterval={43}
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "red",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        >
          No items
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <BVLinearGradient
    colors={
      [
        4294967295,
        3875536895,
        3019898879,
        2164260863,
      ]
    }
    endPoint={
      {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    pointerEvents="none"
    startPoint={
      {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "height": 129,
          "top": 0,
        },
      ]
    }
  />
  <View
    pointerEvents="none"
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "borderBottomColor": "#cecece",
          "borderBottomWidth": 1,
          "top": 129,
        },
      ]
    }
  />
  <BVLinearGradient
    colors={
      [
        2164260863,
        3019898879,
        3875536895,
        4294967295,
      ]
    }
    endPoint={
      {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    pointerEvents="none"
    startPoint={
      {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "bottom": 0,
          "height": 129,
        },
      ]
    }
  />
  <View
    pointerEvents="none"
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "borderTopColor": "#cecece",
          "borderTopWidth": 1,
          "bottom": 129,
        },
      ]
    }
  />
</View>
`;

exports[`Components renders correctly with defaults 2`] = `
<View
  style={
    {
      "alignItems": "center",
      "height": 30,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontFamily": "Georgia",
        "fontSize": 14,
      }
    }
  >
    Item
  </Text>
</View>
`;

exports[`Components renders correctly with gradient disabled 1`] = `
<View
  style={
    {
      "height": 300,
      "width": 300,
    }
  }
>
  <RCTScrollView
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    scrollEventThrottle={20}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    snapToInterval={43}
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "red",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        >
          No items
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 43,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Arial",
              "fontSize": 21.5,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    pointerEvents="none"
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "borderBottomColor": "#cecece",
          "borderBottomWidth": 1,
          "top": 129,
        },
      ]
    }
  />
  <View
    pointerEvents="none"
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "borderTopColor": "#cecece",
          "borderTopWidth": 1,
          "bottom": 129,
        },
      ]
    }
  />
</View>
`;

exports[`Components renders correctly with updated params 1`] = `
<View
  style={
    {
      "height": 400,
      "width": 400,
    }
  }
>
  <RCTScrollView
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    scrollEventThrottle={20}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    snapToInterval={80}
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        >
          Item 1
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        >
          Item 2
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        >
          Item 3
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        >
          Item 4
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        >
          Item 5
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "blue",
              "fontFamily": "Arial",
              "fontSize": 18,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <BVLinearGradient
    colors={
      [
        4294967295,
        3875536895,
        3019898879,
        2164260863,
      ]
    }
    endPoint={
      {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    pointerEvents="none"
    startPoint={
      {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "height": 160,
          "top": 0,
        },
      ]
    }
  />
  <View
    pointerEvents="none"
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "borderBottomColor": "gray",
          "borderBottomWidth": 1,
          "top": 160,
        },
      ]
    }
  />
  <BVLinearGradient
    colors={
      [
        2164260863,
        3019898879,
        3875536895,
        4294967295,
      ]
    }
    endPoint={
      {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    pointerEvents="none"
    startPoint={
      {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "bottom": 0,
          "height": 160,
        },
      ]
    }
  />
  <View
    pointerEvents="none"
    style={
      [
        {
          "position": "absolute",
          "width": "100%",
        },
        {
          "borderTopColor": "gray",
          "borderTopWidth": 1,
          "bottom": 160,
        },
      ]
    }
  />
</View>
`;

exports[`Components renders correctly with updated params 2`] = `
<View
  style={
    {
      "alignItems": "center",
      "height": 40,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      {
        "color": "green",
        "fontFamily": "Arial",
        "fontSize": 22,
      }
    }
  >
    Item label
  </Text>
</View>
`;
